docs/DeveloperGuide.adoc

```diff
@@ -235,6 +235,128 @@ image::CommitActivityDiagram.png[]
+
+[NOTE]
+If the `current` state pointer is at index `history.size() - 1`, i.e. pointing to the latest address book state, then there are no undone states to restore. The `redo` command uses `Model#hasRedo()` to check if this is the case. If so, it will return an error to the user rather than attempting to perform the redo.
+
+Step 5. The user then decides to execute the command `list`. If a command does not modify the data, it will not be stored in the undo history as `UndoHistory` checks for equality with the previous state. Thus, the `history` list remains unchanged.
+
+image::UndoRedoState4.png[]
+
+Step 6. The user executes `clear`, which calls `Model#notifyChange()`. Since the `current` state pointer is not pointing at the end of the `history` list, all address book states after the `current` state pointer will be purged. We designed it this way because it no longer makes sense to redo the `add n/David ...` command. This is the behavior that most modern desktop applications follow.
+
+image::UndoRedoState5.png[]
+
+The following activity diagram summarizes what happens when a user executes a new command:
+
+image::CommitActivityDiagram.png[]
+
+// tag::autocomplete commands[]
+
+=== [Proposed] Autocomplete commands feature
+
+This is a feature which allows you to view all available commands matching the input keyword or letters, eliminating the need to memorize the commands or leave a browser tab open with the UG/DG of this application.
+
+==== Proposed Implementation
+
+The autocomplete mechanism is facilitated by the ‘KeyListener’, which will pass the current character that is being pressed on the keyboard to the trie, a tree-like abstract data type (ADT). The list of matching commands will then be passed back to the ‘Ui’ and displayed via a dropdown box below the user input command box.
+
+For implementation of the trie, please consult ‘google’ or ‘stackoverflow’.
+
+Given below is an example usage scenario and how the autocomplete mechanism behaves at each step.
+
+Step 1: The user launches the application. The current user input command box will be empty.
+
+*insert picture of the ui at startup*
+
+Step 2: The user wants to switch to the customer context for customer-focused commands with the command ‘customer’, however is unsure how to spell the command and types in ‘cus’. The ‘KeyListener’ will pass the characters that have been typed to the trie. The trie will then search for relevant commands and pass them as a list to the ‘Ui’, which will display the relevant results in a dropdown box below the user input command box.
+
+*insert picture of cus being typed in/ class diagram???*
+
+Step 3: The user can now complete the command he/she wants by entering the relevant command shown in the dropdown box.
+
+*insert picture of the ui with the dropdown box*
+
+The following activity diagram summarizes what happens when a user types in letters into the user input command box:
+
+*insert picture of activity diagram*
+
```

Comment:
The DG should focus more on the technical implementation of the feature instead of describing how the feature is used.